cmd/amp-cluster/bootstrap-docker.go

```diff
@@ -0,0 +1,308 @@
+	expected, err := semver.Make(minimumDockerEngineVersion)
+	if err != nil {
+		log.Printf("Minimum version is %s\n", minimumDockerEngineVersion)
+		return false, errors.New("Unable to parse minimum Docker engine version")
+	}
+	observed, err := semver.Make(version)
+	if err != nil {
+		log.Printf("Version is %s\n", version)
+		return false, errors.New("Unable to parse Docker engine version")
+	}
+	return observed.GTE(expected), nil
+}
+
+func (c *clusterClient) isSwarmInit() bool {
+	var out bytes.Buffer
+	var stderr bytes.Buffer
+	var waitStatus syscall.WaitStatus
+	checkCmd := "docker node inspect self"
+	checkCmdSplitted := strings.Split(checkCmd, " ")
+	cmd := exec.Command(checkCmdSplitted[0], checkCmdSplitted[1:]...)
+	cmd.Stdout = &out
+	cmd.Stderr = &stderr
+	err := cmd.Run()
+	if err != nil {
+		if exitError, ok := err.(*exec.ExitError); ok {
+			waitStatus = exitError.Sys().(syscall.WaitStatus)
+			log.Println("Docker swarm mode is not set")
+		} else {
+			log.Printf("Failed to check Swarm status: %s\n", err)
+		}
+		return false
+	}
+	waitStatus = cmd.ProcessState.Sys().(syscall.WaitStatus)
+	if waitStatus != 0 {
+		log.Printf("isSwarmInit: returned %d - %s\n", waitStatus.ExitStatus(), out.String())
+		return false
+	}
+	return true
+}
+
+func (c *clusterClient) joinSwarm() error {
+	if c.dockerClient == nil {
+		defaultHeaders := map[string]string{"User-Agent": "amp-bootstrap"}
+		dclient, err := dockerclient.NewClient(dockerUrl, dockerAPIMinVersion, nil, defaultHeaders)
+		if err != nil {
+			return err
+		}
+		c.dockerClient = dclient
+	}
+	if CreateSwarm == false {
+		if SwarmJoinToken == "" {
+			return errors.New("worker join token not set")
+		}
+		remoteAddresses := []string{SwarmManagerHost + SwarmManagerPort}
+		req := swarm.JoinRequest{ListenAddr: "0.0.0.0:" + SwarmManagerPort, RemoteAddrs: remoteAddresses, JoinToken: SwarmJoinToken}
+
+		return c.dockerClient.SwarmJoin(c.ctx, req)
+	}
+	if Manager == false {
+		return errors.New("won't initialize the Swarm on a worker node")
+	}
+
+	// if multiple IPs, one should be specified
+	ifaces, err := net.Interfaces()
+	var ip net.IP
+	for _, i := range ifaces {
+		addrs, err := i.Addrs()
+		if err != nil {
+			return err
+		}
+		for _, addr := range addrs {
+			switch v := addr.(type) {
+			case *net.IPNet:
+				ip = v.IP
+			case *net.IPAddr:
+				ip = v.IP
+			}
+			// @todo: decide if the IP can be used for the Swarm network (private or public)
+			break
+		}
+	}
+	if ip == nil {
+		return errors.New("No interface available for Swarm init")
+	}
+	req := swarm.InitRequest{ListenAddr: "0.0.0.0:" + SwarmManagerPort, AdvertiseAddr: ip.String()}
+	log.Printf("Swarm init on ip %s\n", ip.String())
+	_, err = c.dockerClient.SwarmInit(c.ctx, req)
+	if err == nil {
+		log.Println("done")
+	}
+	return err
+}
+
+func (c *clusterClient) pullAmpImage(version string) error {
+	image := "appcelerator/amp:" + version
+	_, err := c.dockerClient.ImagePull(c.ctx, image, types.ImagePullOptions{})
```

Comment:
you don't use the reader, so you don't know if pull has error or not during pull